// @tinacms/metrics/dist/index.js
// src/telemetry/telemetry.ts
import { createHash } from "crypto";

// src/telemetry/getId.ts
import { execSync } from "child_process";
function _getProjectIdByGit() {
  try {
    const originBuffer = execSync(
      `git config --local --get remote.origin.url`,
      {
        timeout: 1e3,
        stdio: `pipe`
      }
    );
    return String(originBuffer).trim();
  } catch (_) {
    return null;
  }
}
function getID() {
  return _getProjectIdByGit() || process.env.REPOSITORY_URL || process.cwd();
}

// src/telemetry/getVersion.ts
import fse from "fs-extra";
import { execSync as execSync2 } from "child_process";
import { join } from "path";
function _executeCommand(cmd) {
  try {
    const originBuffer = execSync2(cmd, {
      timeout: 1e3,
      stdio: `pipe`
    });
    return String(originBuffer).trim();
  } catch (_) {
    return null;
  }
}
var _getPack = (rootDir) => {
  let pack = {};
  try {
    const rawJSON = fse.readFileSync(join(rootDir, "package.json")).toString();
    pack = JSON.parse(rawJSON);
  } catch (_e) {
  }
  return pack;
};
var getTinaVersion = () => {
  const pack = _getPack(process.cwd());
  const version = pack?.dependencies?.tinacms;
  return version || "";
};
var getTinaCliVersion = () => {
  const pack = _getPack(process.cwd());
  const version = pack?.devDependencies?.["@tinacms/cli"] || pack?.dependencies?.["@tinacms/cli"];
  return version || "";
};
var getYarnVersion = () => {
  return _executeCommand("yarn -v") || "";
};
var getNpmVersion = () => {
  return _executeCommand("npm -v") || "";
};

// src/telemetry/telemetry.ts
var TINA_METRICS_ENDPOINT = "https://metrics.tina.io/record";
var Telemetry = class {
  //   private config: Conf<Record<string, unknown>>
  projectIDRaw;
  _disabled;
  constructor({ disabled }) {
    this.projectIDRaw = getID();
    const { NO_TELEMETRY } = process.env;
    this._disabled = NO_TELEMETRY === "1" || NO_TELEMETRY === "true" || Boolean(disabled);
  }
  oneWayHash = (payload) => {
    const hash = createHash("sha256");
    hash.update(payload);
    return hash.digest("hex");
  };
  get projectId() {
    return this.oneWayHash(this.projectIDRaw);
  }
  get isDisabled() {
    return this._disabled;
  }
  submitRecord = async ({ event }) => {
    if (this.isDisabled) {
      return;
    }
    try {
      const id = this.projectId;
      const body = {
        partitionKey: id,
        data: {
          anonymousId: id,
          event: event.name,
          properties: {
            ...event,
            nodeVersion: process.version,
            tinaCliVersion: getTinaCliVersion(),
            tinaVersion: getTinaVersion(),
            yarnVersion: getYarnVersion(),
            npmVersion: getNpmVersion(),
            CI: Boolean(process.env.CI)
          }
        }
      };
      await fetch(TINA_METRICS_ENDPOINT, {
        method: "POST",
        body: JSON.stringify(body),
        headers: { "content-type": "application/json" }
      });
    } catch (_e) {
    }
  };
};
export {
  Telemetry
};
